api/billingd/filrewards/filrewards.go

```diff
@@ -0,0 +1,217 @@
+	Key               string          `bson:"key"`
+	AccountType       mdb.AccountType `bson:"account_type"`
+	Event             Event           `bson:"event"`
+	Factor            int             `bson:"factor"`
+	BaseAttoFILReward int             `bson:"base_atto_fil_reward"`
+	CreatedAt         time.Time       `bson:"created_at"`
+}
+
+type FilRewards struct {
+	col          *mongo.Collection
+	rewardsCache map[string]map[Event]struct{}
+	// ToDo: Need some service to send out notifications when rewards are met.
+}
+
+type Config struct {
+	DBURI          string
+	DBName         string
+	CollectionName string
+}
+
+func New(ctx context.Context, config Config) (*FilRewards, error) {
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.DBURI))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.DBName)
+	col := db.Collection(config.CollectionName)
+	_, err = col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys:    bson.D{primitive.E{Key: "key", Value: 1}, primitive.E{Key: "event", Value: 1}},
+			Options: options.Index().SetUnique(true).SetSparse(true),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "key", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "event", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating indexes: %v", err)
+	}
+	// Populate cache.
+	opts := options.Find()
+	cursor, err := col.Find(ctx, bson.M{}, opts)
+	if err != nil {
+		return nil, fmt.Errorf("querying RewardRecords to populate cache: %s", err)
+	}
+	defer cursor.Close(ctx)
+	cache := make(map[string]map[Event]struct{})
+	for cursor.Next(ctx) {
+		var rec RewardRecord
+		if err := cursor.Decode(&rec); err != nil {
+			return nil, fmt.Errorf("decoding RewardRecord while building cache: %v", err)
+		}
+		ensureKeyEventCache(cache, rec.Key)
+		cache[rec.Key][rec.Event] = struct{}{}
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("iterating cursor while building cache: %v", err)
+	}
+	return &FilRewards{
+		col:          col,
+		rewardsCache: cache,
+	}, nil
+}
+
+func (f *FilRewards) ProcessEvent(ctx context.Context, key string, accountType mdb.AccountType, event analytics.Event) (*RewardRecord, error) {
+	rewardEvent := Unspecified
+	switch event {
+	case analytics.KeyAccountCreated:
+		rewardEvent = FirstKeyAccountCreated
+	case analytics.KeyUserCreated:
+		rewardEvent = FirstKeyUserCreated
+	case analytics.OrgCreated:
+		rewardEvent = FirstOrgCreated
+	case analytics.BillingSetup:
+		rewardEvent = InitialBillingSetup
+	case analytics.BucketCreated:
+		rewardEvent = FirstBucketCreated
+	case analytics.BucketArchiveCreated:
+		rewardEvent = FirstBucketArchiveCreated
+	case analytics.MailboxCreated:
+		rewardEvent = FirstMailboxCreated
+	case analytics.ThreadDbCreated:
+		rewardEvent = FirstThreadDbCreated
+	}
+	if rewardEvent == Unspecified {
+		return nil, nil
+	}
```

Comment:
Here is where we map from `analytics.Event` to reward `Event`. If the event is one we're not interested in, just return.

Reply:
Nit: I'd add a comment explaining that `*RewardRecord` may be nil without an error. Might not be obvious to clients and might cause an unwanted nil-pointer dereference.

Reply:
sweet